wallet/src/de/schildbach/wallet/service/ExchangeIntegrationListProvider.kt

```diff
@@ -0,0 +1,186 @@
+import org.dash.wallet.common.integrations.ExchangeIntegration
+import org.dash.wallet.common.integrations.ExchangeIntegrationProvider
+import org.dash.wallet.integrations.coinbase.repository.CoinBaseRepository
+import org.dash.wallet.integrations.coinbase.utils.CoinbaseConfig
+import org.dash.wallet.integrations.uphold.api.UpholdClient
+import org.dash.wallet.integrations.uphold.api.createCardAddress
+import org.dash.wallet.integrations.uphold.api.getAllCards
+import org.dash.wallet.integrations.uphold.api.isAuthenticated
+import org.dash.wallet.integrations.uphold.api.listCardAddress
+import org.dash.wallet.integrations.uphold.utils.UpholdConfig
+import javax.inject.Inject
+
+class ExchangeIntegrationListProvider @Inject constructor(
+    private val coinBaseRepository: CoinBaseRepository,
+    private val coinbaseConfig: CoinbaseConfig,
+    private val upholdConfig: UpholdConfig,
+    private val upholdClient: UpholdClient
+) : ExchangeIntegrationProvider {
+
+    override suspend fun getDepositAddresses(currency: String): List<ExchangeIntegration> {
+        val exchangeIntegrations = arrayListOf<ExchangeIntegration>()
+        // coinbase
+        processCoinbase(currency, exchangeIntegrations)
+        // uphold
+        processUphold(currency, exchangeIntegrations)
+
+        return exchangeIntegrations
+    }
+
+    private suspend fun processCoinbase(
+        currency: String,
+        exchangeIntegrations: ArrayList<ExchangeIntegration>
+    ) {
+        try {
+            if (!lookUpAddress(
+                    "coinbase",
+                    coinbaseConfig,
+                    currency,
+                    R.string.coinbase,
+                    R.drawable.ic_coinbase,
+                    exchangeIntegrations
+                )
+            ) {
+                // determine if we are connected
+                if (coinBaseRepository.isAuthenticated) {
+                    val coinbaseAccount = coinBaseRepository.getUserAccount(currency)
+                    val coinbaseAddress = coinBaseRepository.createAddress(coinbaseAccount.uuid)
+
+                    exchangeIntegrations.add(
+                        ExchangeIntegration(
+                            "coinbase",
+                            coinBaseRepository.isAuthenticated,
+                            coinbaseAddress.unwrap(),
+                            currency,
+                            R.string.coinbase,
+                            R.drawable.ic_coinbase
+                        )
+                    )
+                } else {
+                    exchangeIntegrations.add(
+                        ExchangeIntegration(
+                            "coinbase",
+                            false,
+                            null,
+                            currency,
+                            R.string.coinbase,
+                            R.drawable.ic_coinbase
+                        )
+                    )
+                }
+            }
```

Comment:
The idea is to look in the cache (the config for that exchange), then query accounts and then an address.  The exchange repository or client will save any results to the cache.